### ISSUE TITLE
How (and if?) do we model relations between contacts/people and their roles?

### ISSUE THREAD

**Issue by wrygiel** [erasmus-without-paper/ewp-specs-api-institutions#5]
As @kaiqu explained [here](https://github.com/erasmus-without-paper/ewp-wp3-data-model/issues/5#issuecomment-256578595), we might need relations between [`<contact>`s](https://github.com/erasmus-without-paper/ewp-specs-api-institutions/blob/v0.1.0/response.xsd#L206) and IIAs.

In version `0.1.0` of the Institutions API schema, this is dealt with via the [`<iia-id>`](https://github.com/erasmus-without-paper/ewp-specs-api-institutions/blob/v0.1.0/response.xsd#L359) sequence, but - if I understood this correctly - this is not enough. We need to able to assign iia-ids for every role of every contact.

@kaiqu, did I understand this correctly?

**Comment by kaiqu** [comment]
> As @kaiqu explained here, we might need relations between "contact"s and IIAs.
> 
> In version 0.1.0 of the Institutions API schema, this is dealt with via the "iia-id" sequence, but - if I understood this correctly - this is not enough. We need to able to assign iia-ids for every role of every contact.
> 
> @kaiqu, did I understand this correctly?

First, some data model terminology:
- Coordinators are Persons with Roles
- Contacts are Coordinators for Mobilities

A Coordinator belongs to an Institution/Org Unit, and can (in the next version) be connected to 0..\* IIAs. I think the latter connections belong in the IIAs API.

BTW, I suggest we use "quotes" (or something) instead of XML <brackets> for elements, since the latter convention aren't displayed in quoted text.

**Comment by wrygiel** [comment]
> Contacts are Coordinators for Mobilities

Oops, I have used a different terminology [in my XSD](https://github.com/erasmus-without-paper/ewp-specs-api-institutions/blob/v0.1.0/response.xsd#L206). Are these terms defined in WP2's dictionary?

My "contact" element is not necessarily a person. It can be, for example, an office (a phone number, or an email address, shared by a couple of people). Do you think it's a wrong assumption on my part? Do you think that we can require all coordinators to be _specific_ persons?

> A Coordinator belongs to an Institution/Org Unit, and can (in the next version) be connected to 0..\* IIAs. I think the latter connections belong in the IIAs API.

I would rather assume that coordinators are part of factsheets (and factsheets are served in Institutions API).

In my mind, these two entities are divided quite clearly, according to the following:
- Factsheets contain information which is public, and can change in time without the need of signing the IIA again.
- IIAs contain information which is _not necessarily_ public, and - if it changes - it requires the agreement to be signed again.

**Comment by wrygiel** [comment]
> My "contact" element is not necessarily a person.

For example, see [here](https://cloud.githubusercontent.com/assets/2168535/19776737/9be7d368-9c75-11e6-8e13-4d4cf5409768.png) ([source](https://studyabroad.hh.umu.se/utland/factsheets/NHH-%2813_14%29_20140926125346.pdf)), or [here](https://cloud.githubusercontent.com/assets/2168535/19776790/c5e91c8a-9c75-11e6-9357-d6f6e87f9140.png) ([source](https://studyabroad.hh.umu.se/utland/factsheets/BI-%2816_17%29_20160129132607.pdf)).

([more factsheet examples](https://studyabroad.hh.umu.se/utland/valj_universitylist.php?sort=country))

**Comment by kaiqu** [comment]
> > Contacts are Coordinators for Mobilities
> 
> Oops, I have used a different terminology in my XSD. Are these terms defined in WP2's dictionary?

No - the task of aligning the model with the dictionary was assigned a lower priority, to be done after the structure is finalized.

> My "contact" element is not necessarily a person. It can be, for example, an office (a phone number, or an email address, shared by a couple of people). Do you think it's a wrong assumption on my part? Do you think that we can require all coordinators to be specific persons?

This is a matter of terminology, rather than structure (it was probably not a good idea to delay the above-mentioned aligning with the dictionary, since it's important to know exactly what we are talking about when designing the structure). In the model, there is a sharp delineation between two types of contact information:
- For _people_ it's defined in the Coordinator entity
- For _functions_ it's defined in Inst/Org Unit attributes

I've never been quite happy with the latter, so now might be the time to change it.
In my mind, the term "coordinator" denotes a person, while "contact" is more general.
I propose the following:
- An Inst/Org Unit has Contacts and Coordinators
- A Contact is basically a URL/Email/Phone with a Role
- Coordinator stays the same (a Person with a Role, independent of IIAs)
- The present Contact entity (Coordinator for a Mobility) is renamed to Mobility Coordinator

> > A Coordinator belongs to an Institution/Org Unit, and can (in the next version) be connected to 0..\* IIAs. I think the latter connections belong in the IIAs API.
> 
> I would rather assume that coordinators are part of factsheets (and factsheets are served in Institutions API).

The Coordinators themselves, yes. I was referring to their _connections to IIAs_ - i.e. a Coordinator having Role for an IIA (denoted in the coming version by a N:M relationship).

> In my mind, these two entities are divided quite clearly, according to the following:
> 
> Factsheets contain information which is public, and can change in time without the need of signing the IIA again.
> IIAs contain information which is not necessarily public, and - if it changes - it requires the agreement to be signed again.

With that distinction, I guess assigning a new Coordinator to an IIA doesn't necessitate re-signing, so then I agree that the assignment of Coordinators to IIAs belong in the Institutions/Departments APIs.

**Comment by kaiqu** [comment]
> My "contact" element is not necessarily a person.
> 
> For example, see here (source), or here (source).

The Contact term is discussed in my previous message, but I notice that the above examples seem to indicate more Roles than the ones mentioned so far.

> (more factsheet examples)

I looked at some of these in the beginning of the modeling process, but it would probably be a good idea to go through some more fact sheets at this point to 1) check that the model hasn't lost any information and 2) complete code sets such as Role...

**Comment by wrygiel** [comment]
> (...) two types of contact information: For people it's defined in the Coordinator entity

The term "coordinator" seems to fit our current set of roles. But I can imagine that as the list of roles grows, it might become very awkward, at some point. Intuitively, "coordinating something" seems to be a subset of all possible roles (and we can expect for this set to be growing).

BTW - our model image is becoming a bit tangled up. Do we need to model all entities on a single image? Perhaps it would be easier for the readers to split it up, for example, by the API?

Before I tackle your other comments - let's think about minimum requirements again, for a moment. The more I think on this, the more I wonder if we really _need_ to model all these roles. Since EWP allows communication between whole systems, then perhaps it's up to the particular system developer to know _who_ is supposed to do _what_ in _which_ context, and the partner doesn't need to know that. If this is true, then do we need roles in our model at all? Or, perhaps, we need only a small subset?

I would start with the smallest set of features required to meet our use cases. We can always add more features as we go.

**Comment by wrygiel** [comment]
Since it's somewhat unclear to me at this point, I have moved `<role-id>` and other related elements to a separate feature branch. (So that the `master` branch stays "clean".)

**Comment by janinamincer-daszkiewicz** [comment]
> No - the task of aligning the model with the dictionary was assigned a lower priority, to be done after the structure is finalized.

It is a good assumpiton if we talk about technical dictionary prepared by WP3, _not_ terminology prepared by WP2. That should be taken care of in the first place.

**Comment by janinamincer-daszkiewicz** [comment]
> I looked at some of these in the beginning of the modeling process, but it would probably be a good idea to go through some more fact sheets at this point

You will find a substantial set gathered once by the partners in the Mobility project:
https://www.usos.edu.pl/Mobility/InfoSheets/

In fact any proposal for the model should first be checked against this set of fact sheets.

**Comment by janinamincer-daszkiewicz** [comment]
> I would start with the smallest set of features required to meet our use cases. We can always add more features as we go.

I agree.

**Comment by kaiqu** [comment]
> The term "coordinator" seems to fit our current set of roles. But I can imagine that as the list of roles grows, it might become very awkward, at some point. Intuitively, "coordinating something" seems to be a subset of all possible roles (and we can expect for this set to be growing).

Yes.

> BTW - our model image is becoming a bit tangled up. Do we need to model all entities on a single image? Perhaps it would be easier for the readers to split it up, for example, by the API?

Splitting up by API is not really possible, as many entities are used in more than one API. I guess it could be split up into some kind of "business areas", associating entities with a "primary" API:

**Organization**
- Institution/Org Unit
- Coordinator
- HEI Contact Info
- Person
- Person Contact Info
- Contact Type
- Learning Opportunity Specification
- Learning Opportunity Instance
- Academic Term/Year
- Result Distribution

**IIA**
- IIA
- IIA Partner
- Cooperation Condition

**Mobility**
- Mobility
- Learning Agreement
- LA Component

There are basically three ways to do this (ranging from the easiest to the most work):
1. Assigning three _colors_ to the entities
2. Blowing the model up to A3 format
3. Distributing the entities on 2 pages (the last two should fit one the same page)

I would like to try option 1 first, if we can assume everyone to have access to color printers?

> Before I tackle your other comments - let's think about minimum requirements again, for a moment. The more I think on this, the more I wonder if we really need to model all these roles. Since EWP allows communication between whole systems, then perhaps it's up to the particular system developer to know who is supposed to do what in which context, and the partner doesn't need to know that. If this is true, then do we need roles in our model at all? Or, perhaps, we need only a small subset?

I assume contact information is only visual (if it was to be used programmatically in a different way for different Roles, the Roles would definitely be necessary). The only reason, then, for having defined Roles would be to be able to programmatically display contact information in a standard way for those Roles (and possibly to determine which Roles are assigned at a HEI).

> I would start with the smallest set of features required to meet our use cases. We can always add more features as we go.

I agree. If Roles are deemed to be outside the smallest set, I'll remove them.

**Comment by kaiqu** [comment]
> > the task of aligning the model with the dictionary was assigned a lower priority, to be done after the structure is finalized.
> 
> It is a good assumpiton if we talk about technical dictionary prepared by WP3, not terminology prepared by WP2. That should be taken care of in the first place.

Aha. Ok, I will do that before releasing the next version of the model.

**Comment by kaiqu** [comment]
> You will find a substantial set gathered once by the partners in the Mobility project:
> https://www.usos.edu.pl/Mobility/InfoSheets/
> 
> In fact any proposal for the model should first be checked against this set of fact sheets.

There are more than 70 documents here (counting the ones in the subdirectories) - is it possible to identify a key subset of them, for me to check?

**Comment by janinamincer-daszkiewicz** [comment]
I would have to look at them myself ;)
I did but it was a couple of years ago.
Just pick up 10 randomly.

**Comment by janinamincer-daszkiewicz** [comment]
It might be of some use to have a look at businessCardT in:
https://www.usos.edu.pl/Mobility/Documentation2/default.html

**Comment by kaiqu** [comment]
> It might be of some use to have a look at businessCardT in:
> https://www.usos.edu.pl/Mobility/Documentation2/default.html

Hm... Leaving out userDefinedBusinessCardSectionT and userDefinedSectionItemT (a generic name/value structure), the remaining components seem to be modeled by the current version:

organizationBCSection is covered by:
- Institution/Org Unit
- HEI Contact Info

iroOfficeBCSection is covered by:
- Institution/Org Unit
- HEI Contact Info
- Coordinator
- Person
- Person Contact Info

So, unless there is need for such a generic name/value structure, the model seems sufficient in this respect.

**Comment by kaiqu** [comment]
> > It is a good assumpiton if we talk about technical dictionary prepared by WP3, not terminology prepared by WP2. That should be taken care of in the first place.
> 
> Aha. Ok, I will do that before releasing the next version of the model.

I have produced a version of the WP2 Dictionary (data_dictionary.xslx), where I map from the Field Name to the current model, expressed by two added columns "WP3 - Attribute" and "WP3 - Entity". There are some questions which should be cleared up - should I commit the spreadsheet (possibly to a branch), or would this be considered unnecessary clutter in the repo?

**Comment by kaiqu** [comment]
> I have produced a version of the WP2 Dictionary (data_dictionary.xslx), where I map from the Field Name to the current model, expressed by two added columns "WP3 - Attribute" and "WP3 - Entity". There are some questions which should be cleared up - should I commit the spreadsheet (possibly to a branch), or would this be considered unnecessary clutter in the repo?

I tried pushing the commented file to a new branch, but the server returns "permission denied" / HTTP 403. Should I have write access to this repo, or is it better to send the commented file to the technical mailing list?

@wrygiel @erasmus-without-paper/wp2

**Comment by janinamincer-daszkiewicz** [comment]
> So, unless there is need for such a generic name/value structure, the model seems sufficient in this respect.

In my opinion there is a need for such generic section since it gives the possibility to send any information which might seem relevent by sending HEI. Our old Mobility client and server is in operation, you can see some real examples at:
http://mobility.usos.edu.pl/mobility-server-transport/
These are business cards prepared by partner institutions.

**Comment by kaiqu** [comment]
> > So, unless there is need for such a generic name/value structure, the model seems sufficient in this respect.
> 
> In my opinion there is a need for such generic section since it gives the possibility to send any information which might seem relevent by sending HEI. Our old Mobility client and server is in operation, you can see some real examples at:
> http://mobility.usos.edu.pl/mobility-server-transport/
> These are business cards prepared by partner institutions.

I think this is solved by the Info attribute in Inst/Org Unit in the latest version of the model. It represents 0..\* Info elements with Headers for display/grouping (as described in my message to the technical list yesterday). I propose to replace the predefined Role IDs with this.

**Comment by janinamincer-daszkiewicz** [comment]
Is this already reflected in the XSD or XML example?

**Comment by janinamincer-daszkiewicz** [comment]
Are you sure that all posted Business Cards (Fact Sheets) from the Mobility server can be easily described by the model? Could we see the XML examples showing this? That would be real proof of concept.

**Comment by janinamincer-daszkiewicz** [comment]
What you see at
http://mobility.usos.edu.pl/mobility-server-transport/
is posted automatically by the server after API invocation of the client. XML contains header (title) section which helps to structure information in a way readable for humans. We might even post such Facts Sheets of EWP partners at the registry web page (if we decide to have one and if there comes an attribute 'public' with the invocation).

**Comment by kaiqu** [comment]
> Are you sure that all posted Business Cards (Fact Sheets) from the Mobility servrr can be easily described by the model? Could we see the XML examples showing this? That would be real proof of concept.

I am working on it.

> What you see at
> http://mobility.usos.edu.pl/mobility-server-transport/
> is posted automatically by the server after API invocation of the client. XML contains header (title) section which helps to structure information in a way readable for humans.

This fits the Info structure perfectly (header + content repeated), which was the result of examining quite a few fact sheets (including the one from UW). It also fits pretty well with what @wrygiel created in a branch (except he called the structure "embedded-mobility-factsheet" and made it embeddable/recursive).

Also, since there doesn't seem to be any need for programmatic processing of predefined Role IDs, I propose to replace them with headers. That way, we get the header/content structure for all kinds of repeating information connected to Inst/Org Unit. (Coming in the next model version)

**Comment by wrygiel** [comment]
> I tried pushing the commented file to a new branch, but the server returns "permission denied" / HTTP 403.

In general, only WP2 is allowed to change things in `ewp-wp2-*` repositories, because they are responsible for their contents. If you would like to suggest changes, the "GitHub way" of doing this is to create a **fork** (your local copy of the repository), and then make a **pull request**. Then, you assign people to this pull request to review your comments.

**Comment by wrygiel** [comment]
BTW, @kaiqu, I see we are working on one API in parallel. I hope it's not too confusing for you :) I also have some local changes for Organizational Units API, but I'm not sure if will be able to push them before I leave for my holidays.

**Comment by kaiqu** [comment]
> In general, only WP2 is allowed to change things in ewp-wp2-\* repositories, because they are responsible for their contents. If you would like to suggest changes, the "GitHub way" of doing this is to create a fork (your local copy of the repository), and then make a pull request. Then, you assign people to this pull request to review your comments.

Sorry about the GitHub newbie questions, but I have created the fork and pull request, but I don't see how to assign people to it?

**Comment by kaiqu** [comment]
> BTW, @kaiqu, I see we are working on one API in parallel. I hope it's not too confusing for you :) I also have some local changes for Organizational Units API, but I'm not sure if will be able to push them before I leave for my holidays.

No problem :-)
I'm waiting with the Org Units API until the Institutions API is finished, since they will be very similar.

**Comment by wrygiel** [comment]
> I have created the fork and pull request, but I don't see how to assign people to it

It seems that you'd need to have write-permissions to assign people directly. So, lets use @mention tags instead.

**Comment by wrygiel** [comment]
> As @kaiqu explained [here](https://github.com/erasmus-without-paper/ewp-wp3-data-model/issues/5#issuecomment-256578595), we might need relations between [`<contact>`s](https://github.com/erasmus-without-paper/ewp-specs-api-institutions/blob/v0.1.0/response.xsd#L206) and IIAs.

In the latest IIAs API draft (version 0.1.0), we have dealt with this problem by allowing `<contact>` elements in various places of the IIA element tree. Each such contact can have a `<role-description>`. Contacts with very specific roles have separate elements to distinguish them. We will see if this proves enough for our purposes (new, more specific issues might arise in the future).

BTW, this also means that some data that we initially planned to put in the Institutions API, will now be placed in the IIAs API (as explained [here](https://github.com/erasmus-without-paper/ewp-specs-api-iias/tree/v0.1.0#where-are-the-fact-sheets)).
